packages/fulfillment/src/services/fulfillment-module-service.ts

```diff
@@ -1074,6 +1207,82 @@ export default class FulfillmentModuleService<
       : updatedShippingOptionRules[0]
   }
 
+  @InjectTransactionManager("baseRepository_")
+  async updateFulfillment(
+    id: string,
+    data: FulfillmentTypes.UpdateFulfillmentDTO,
+    @MedusaContext() sharedContext: Context = {}
+  ): Promise<FulfillmentTypes.FulfillmentDTO> {
+    const fulfillment = await this.fulfillmentService_.update(
+      { id, ...data },
+      sharedContext
+    )
+
+    const serialized =
+      await this.baseRepository_.serialize<FulfillmentTypes.FulfillmentDTO>(
+        fulfillment,
+        {
+          populate: true,
+        }
+      )
+
+    return Array.isArray(serialized) ? serialized[0] : serialized
+  }
+
+  @InjectManager("baseRepository_")
+  async cancelFulfillment(
+    id: string,
+    @MedusaContext() sharedContext: Context = {}
+  ): Promise<FulfillmentDTO> {
+    const canceledAt = new Date()
+
+    let fulfillment = await this.fulfillmentService_.retrieve(
+      id,
+      {},
+      sharedContext
+    )
+
+    if (fulfillment.shipped_at) {
```

Comment:
Let's move all validations to a separate function, it's a bit easier to read when we separate validation from other business logic